src/main/scala/edu/ie3/simona/agent/grid/PowerFlowSupport.scala

```diff
@@ -77,93 +82,195 @@
+        .collect {
+          case Some(providePowerMessage: ProvidePowerMessage) =>
+            providePowerMessage
+          case Some(message) =>
+            throw new RuntimeException(
+              s"Received message $message which cannot processed here!"
+            )
+          case None =>
+            throw new RuntimeException(
+              s"Did not receive all power values I expected @ node $nodeId. This is a fatal and should never happen!"
+            )
         }
+        .foldLeft(
+          (
+            Quantities.getQuantity(0, powerUnit),
+            Quantities.getQuantity(0, powerUnit)
+          )
+        )((pqSum, powerMessage) => {
+          (pqSum._1.add(powerMessage.p), pqSum._2.add(powerMessage.q))
+        })
 
-      val presetNodeData =
-        PresetData(nodeIdx, nodeType, apparentPower, targetVoltageInPu.abs)
+      new Complex(
+        refSystem.pInPu(p).getValue.doubleValue(),
+        refSystem.qInPu(q).getValue.doubleValue()
+      )
+    case None => new Complex(0, 0)
+  }
 
-      operatingPoint :+ presetNodeData
+  /** Determine a target voltage if no received values are meant to be taken
+    *
+    * @param nodeModel
+    *   The nodal model
+    * @param ignoreTargetVoltage
+    *   Whether or not to ignore the nodal target voltage provided by input
+    *   model
+    * @return
+    *   A complex nodal target voltage
+    */
+  private def nodalTargetVoltage(
+      nodeModel: NodeModel,
+      ignoreTargetVoltage: Boolean
+  ): Complex = Complex.one * (if (!ignoreTargetVoltage)
+                                nodeModel.vTarget
+                                  .to(PowerSystemUnits.PU)
+                                  .getValue
+                                  .doubleValue()
+                              else 1.0)
 
-    })
+  /** Determine node type, target voltage and state data (dependent of the node
+    * type)
+    *
+    * @param nodeModel
+    *   Complete node model
+    * @param nodeIndex
+    *   Index of the node
+    * @param receivedValuesStore
+    *   Store of received data
+    * @param transformers2w
+    *   Set of two winding transformer models
+    * @param transformers3w
+    *   Set of three winding transformer models
+    * @param residualPower
+    *   Relative nodal residual power
+    * @param stateData
+    *   Current collection of state data
+    * @param refSystem
+    *   Reference system to use
+    * @param targetVoltageFromReceivedData
+    *   Whether or not to take target voltage from received data
+    * @param ignoreTargetVoltage
+    *   Whether or not to neglect the target voltage given in node input
+    * @return
+    */
+  private def nodeTypeTargetVoltageStateData(
+      nodeModel: NodeModel,
+      nodeIndex: Int,
+      receivedValuesStore: ReceivedValuesStore,
+      transformers2w: Set[TransformerModel],
+      transformers3w: Set[Transformer3wModel],
+      residualPower: Complex,
+      stateData: Array[StateData],
```

Comment:
I'd leave out this parameter here, as it's pretty much only used when adding to the array (line 261) or returned as-is (line 272). Instead, I would change the return type of this method to `(NodeType, Complex, Option[StateData])`, so that the optional stateData is only added to the array (in `composeOperatingPoint`) if it's present.

(this then also goes towards fixing one of the two new Codacy issues (having more than 8 params))